examples/adjoints/adjoints_ex7/SigmaPhysics.C

```diff
@@ -0,0 +1,194 @@
+
+#include "libmesh/fe_base.h"
+#include "libmesh/fe_interface.h"
+#include "libmesh/fem_context.h"
+#include "libmesh/mesh.h"
+#include "libmesh/quadrature.h"
+#include "libmesh/string_to_enum.h"
+#include "libmesh/system.h"
+#include "libmesh/equation_systems.h"
+#include "libmesh/zero_function.h"
+#include "libmesh/const_function.h"
+#include "libmesh/dirichlet_boundaries.h"
+#include "libmesh/dof_map.h"
+#include "libmesh/numeric_vector.h"
+#include "heatsystem.h"
+
+void SigmaPhysics::init_data (System & sys)
+{
+  T_var = dynamic_cast<FEMSystem&>(sys).add_variable ("T", static_cast<Order>(1), LAGRANGE);
+
+  GetPot infile("heat.in");
+  _k = infile("k", 1.0);
+  _analytic_jacobians = infile("analytic_jacobians", true);
+
+  // The temperature is evolving, with a first order time derivative
+  this->time_evolving(T_var, 1);
+
+  const boundary_id_type all_ids[4] = {0, 1, 2, 3};
+  std::set<boundary_id_type> all_bdys(all_ids, all_ids+(2*2));
+
+  std::vector<unsigned int> T_only(1, T_var);
+
+  ZeroFunction<Number> zero;
+
+  ConstFunction<Number> one(1.0);
+
+  // Most DirichletBoundary users will want to supply a "locally
+  // indexed" functor
+  sys.get_dof_map().add_dirichlet_boundary(DirichletBoundary (all_bdys, T_only, &one));
+
+  // Set adjoint boundary condition z(1,y) = y * (1 - y)
+  sys.get_dof_map().add_adjoint_dirichlet_boundary(DirichletBoundary (all_bdys, T_only, &zero), 0);
+  sys.get_dof_map().add_adjoint_dirichlet_boundary(DirichletBoundary (all_bdys, T_only, &zero), 1);
+
+}
+
+void SigmaPhysics::init_context(DiffContext & context)
+{
+  FEMContext & c = cast_ref<FEMContext &>(context);
+
+  FEBase * elem_fe = nullptr;
+  c.get_element_fe(0, elem_fe);
+
+  // Now make sure we have requested all the data
+  // we need to build the linear system.
+  elem_fe->get_JxW();
+  elem_fe->get_dphi();
+  elem_fe->get_phi();
+  elem_fe->get_xyz();
+
+  // Don't waste time on side computations for T
+  FEBase * side_fe = nullptr;
+  c.get_side_fe(0, side_fe);
+  side_fe->get_nothing();
+
+  // We'll have a more automatic solution to preparing adjoint
+  // solutions for time integration, eventually...
+  if (c.is_adjoint())
+    {
+      // A reference to the system context is built with
+      const System & sys = c.get_system();
+
+      // Get a pointer to the adjoint solution vector
+      NumericVector<Number> & adjoint_solution0 =
+        const_cast<System &>(sys).get_adjoint_solution(0);
+
+      // Add this adjoint solution to the vectors that diff context should localize
+      c.add_localized_vector(adjoint_solution0, sys);
+
+      // Get a pointer to the adjoint solution vector
+      NumericVector<Number> & adjoint_solution1 =
+        const_cast<System &>(sys).get_adjoint_solution(1);
+
+      // Add this adjoint solution to the vectors that diff context should localize
+      c.add_localized_vector(adjoint_solution1, sys);
+    }
+
+}
+
+#define optassert(X) {if (!(X)) libmesh_error_msg("Assertion " #X " failed.");}
```

Comment:
Is this getting used anywhere?

Reply:
Its a workaround for Fedora OS issue IIRC.

Reply:
Okay, I can find it in my svn logs now.  The comment later in the file is "Workaround for weird FC6 bug".  That's Fedora Core 6, released 2006.  The reason I had to go back to svn logs is because the compiler bug nearly predated git!

We don't support anything without C++11 anymore, much less anything with such weird bugs.  Remove it.